gap/semigroups/semitrans.gi

```diff
@@ -283,49 +283,72 @@ SEMIGROUPS.LargestElementRClass := function(R)
   return SEMIGROUPS.ElementRClass(R, true);
 end;
 
-# stop_on_isolated_pair:
-#   if true, this function returns false if there is an isolated pair-vertex
-# TODO should this be a Digraph??
-# FIXME make this a digraph
-SEMIGROUPS.GraphOfRightActionOnPairs := function(gens, n, stop_on_isolated_pair)
-  local nrgens, nrpairs, PairNumber, NumberPair, in_nbs, labels, pair,
-  isolated_pair, act, range, i, j;
-  nrgens     := Length(gens);
-  nrpairs    := Binomial(n, 2);
+InstallMethod(DigraphOfActionOnPairs,
+"for a transformation semigroup",
+[IsTransformationSemigroup],
+function(S)
+  return DigraphOfActionOnPairs(GeneratorsOfSemigroup(S),
+                                DegreeOfTransformationSemigroup(S));
+end);
+
+InstallMethod(DigraphOfActionOnPairs,
+"for a transformation collection",
+[IsTransformationCollection],
+function(coll)
+  return DigraphOfActionOnPairs(coll, DegreeOfTransformationCollection(coll));
+end);
+
+InstallMethod(DigraphOfActionOnPairs,
+"for a transformation collection and pos int",
+[IsTransformationCollection, IsInt],
+function(coll, n)
+  local PairNumber, NumberPair, nr, nrpairs, out_nbs, inn_nbs, label, pair,
+  range, gr, i, j;
+
+  if n < 0 then
+    ErrorNoReturn("Semigroups: DigraphOfActionOnPairs: usage,\n",
+                  "the second argument <n> must be non-negative,");
+  elif n <= 1 then
+    return EmptyDigraph(n);
+  fi;
+
   PairNumber := Concatenation([1 .. n], Combinations([1 .. n], 2));
-  # Currently assume <x> is sorted and is a valid combination of [1 .. n]
   NumberPair := function(n, x)
+    # This function assumes that <x> is a sorted subset of [1 .. n]
     if Length(x) = 1 then
       return x[1];
     fi;
     return n + Binomial(n, 2) - Binomial(n + 1 - x[1], 2) + x[2] - x[1];
   end;
 
-  in_nbs := List([1 .. n + nrpairs], x -> []);
-  labels := List([1 .. n + nrpairs], x -> []);
-  for i in [(n + 1) .. (n + nrpairs)] do
+  if IsSemigroup(coll) then
+    coll := GeneratorsOfSemigroup(coll);
+  fi;
+
+  nr         := Length(coll);
+  nrpairs    := Binomial(n, 2);
+  out_nbs := List([1 .. n + nrpairs], x -> []);
+  inn_nbs := List([1 .. n + nrpairs], x -> []);
+  label   := List([1 .. n + nrpairs], x -> []);
+
+  for i in [n + 1 .. n + nrpairs] do
     pair := PairNumber[i];
-    isolated_pair := true;
-    for j in [1 .. nrgens] do
-      act := OnSets(pair, gens[j]);
-      range := NumberPair(n, act);
-      Add(in_nbs[range], i);
-      Add(labels[range], j);
-      if range <> i and isolated_pair then
-        isolated_pair := false;
+    for j in [1 .. nr] do
+      range := NumberPair(n, OnSets(pair, coll[j]));
+      if i <> range and not i in inn_nbs[range] then
+        Add(inn_nbs[range], i);
+        Add(out_nbs[i], range);
+        Add(label[i], j);
```

Comment:
If I didn't need to keep track of edge labels, I would do this. However I am labelling edges which would make this tricky.

The ordering of `label[i]` corresponds with the ordering of `out_nbs[i]`. For example, there is an edge from `i` to `out_nbs[i][2]` with label `label[i][2]`. Suppose that I want to store the edge `i -> range` with label `j` by doing `AddSet(out_nbs[i], range)`. Then what is the proper position of `label[i]` in which to insert `j`? Or did the edge `i -> range` already exist, meaning that its label doesn't need to be stored?

Do you see a way around these problems?